cli/src/action/login/oauth/callback.rs

```diff
@@ -0,0 +1,107 @@
+// Copyright 2018-2020 Cargill Incorporated
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+
+use tiny_http::{Header, Method, Request, Response, Server};
+
+use super::error::OAuth2CallbackError;
+
+pub struct OAuth2Callback {
+    port: u16,
+    join_handle: std::thread::JoinHandle<Result<String, OAuth2CallbackError>>,
+}
+
+impl OAuth2Callback {
+    pub fn new() -> Result<Self, OAuth2CallbackError> {
+        let server =
+            Server::http("127.0.0.1:0").map_err(|err| OAuth2CallbackError(err.to_string()))?;
+
+        let port = server.server_addr().port();
+
+        let join_handle = std::thread::Builder::new()
+            .name("Thread-OAuth2Callback".into())
+            .spawn(move || Self::run_callback_server(server))
+            .map_err(|err| OAuth2CallbackError(err.to_string()))?;
+
+        Ok(Self { join_handle, port })
+    }
+
+    pub fn address(&self) -> String {
+        format!("localhost:{}", self.port)
+    }
+
+    pub fn receive_authorization_code(self) -> Result<String, OAuth2CallbackError> {
```

Comment:
Can we rename this to `wait_for_authorization_code` or something similar to indicate that this is a blocking method?